metric/id/m3/id.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package m3 describes m3 metric id information.
+//
+// Each M3 metric id contains a metric name and a set of tag pairs. In particular,
+// it conforms to the following format:
+//
+// m3+<metric_name>+tagName1=tagValue1,tagName2=tagValue2,...tagNameN=tagValueN,
+//
+// Where the tag names are sorted alphabetically in ascending order.
+//
+// An example m3 metrid id is as follows:
+// m3+response_code+env=bar,service=foo,status=404,type=counters
+package m3
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3metrics/metric/id"
+)
+
+const (
+	componentSplitter = '+'
+	tagNameSplitter   = '='
+	tagPairSplitter   = ','
+)
+
+var (
+	errInvalidM3Metric = fmt.Errorf("invalid m3 metric")
+	m3Prefix           = []byte("m3+")
+	nameTagName        = []byte("name")
+	rollupTagPair      = id.TagPair{
+		Name:  []byte("m3_rollup"),
+		Value: []byte("true"),
+	}
+)
+
+// NewRollupID generates a new rollup id given the new metric name
+// and a list of tag pairs. Note that tagPairs are mutated in place.
+func NewRollupID(name []byte, tagPairs []id.TagPair) []byte {
+	var buf bytes.Buffer
+
+	// Adding rollup tag pair to the list of tag pairs.
+	tagPairs = append(tagPairs, rollupTagPair)
+	sort.Sort(id.TagPairsByNameAsc(tagPairs))
```

Comment:
It so happens that it is sorted atm but doesn't have to be. Seems a bit weird to require the caller pass in sorted tag pairs. Also because these are used for rollup tags, so chances are that the number of tag pairs is gonna be small and the perf diff between the two approaches is going to be immaterial.